examples/transfer-collection.rs

```diff
@@ -0,0 +1,126 @@
+/// shared state you'd want to couple with network operations.
+struct Node {
+    store: Store,
+    /// Router with the blobs protocol registered, to accept blobs requests.
+    /// We can always get the endpoint with router.endpoint()
+    router: Router,
+}
+
+impl Node {
+    async fn new() -> Result<Self> {
+        let endpoint = Endpoint::builder().bind().await?;
+
+        let store = MemStore::new();
+
+        // this BlobsProtocol accepts connections from other nodes and serves blobs from the store
+        // we pass None to skip subscribing to request events
+        let blobs = BlobsProtocol::new(&store, endpoint.clone(), None);
+        // Routers group one or more protocols together to accept connections from other nodes,
+        // here we're only using one, but could add more in a real world use case as needed
+        let router = Router::builder(endpoint)
+            .accept(iroh_blobs::ALPN, blobs)
+            .spawn();
+
+        Ok(Self {
+            store: store.into(),
+            router,
+        })
+    }
+
+    // get address of this node. Has the side effect of waiting for the node
+    // to be online & ready to accept connections
+    async fn node_addr(&self) -> Result<NodeAddr> {
+        let addr = self.router.endpoint().node_addr().initialized().await;
+        Ok(addr)
+    }
+
+    async fn list_hashes(&self) -> Result<Vec<Hash>> {
+        self.store
+            .blobs()
+            .list()
+            .hashes()
+            .await
+            .context("Failed to list hashes")
+    }
+
+    /// creates a collection from a given set of named blobs, adds it to the local store
+    /// and returns the hash of the collection.
+    async fn create_collection(&self, named_blobs: Vec<(&str, Vec<u8>)>) -> Result<Hash> {
+        let mut collection_items: HashMap<&str, TempTag> = HashMap::new();
+
+        let tx = self.store.batch().await?;
+        for (name, data) in named_blobs {
+            let tmp_tag = tx.add_slice(data).await?;
+            collection_items.insert(name, tmp_tag);
+        }
+
+        let collection_items = collection_items
+            .iter()
+            .map(|(name, tag)| {
+                let hash = tag.hash().clone();
+                (name.to_string(), hash)
+            })
+            .collect::<Vec<_>>();
+
+        let collection = Collection::from_iter(collection_items);
+
+        let collection = collection.store(&self.store).await?;
+        let hash = collection.hash().clone();
+        self.store.tags().create(collection).await?;
+        Ok(hash)
+    }
+
+    /// retrive an entire collection from a given hash and provider
+    async fn get_collection(&self, hash: Hash, provider: NodeAddr) -> Result<()> {
+        self.router.endpoint().add_node_addr(provider.clone())?;
+        let req = HashAndFormat::hash_seq(hash);
+        let addrs = Shuffled::new(vec![provider.node_id]);
+        self.store
+            .downloader(self.router.endpoint())
```

Comment:
I'm super confused about why I have to re-pass the endpoint to the store here to create a downloader. I guess I'm supposed to make a downloader & store it on the `Node` struct I've defined here?

It generally feels like a higher-level API is missing that just falls back to a `DefaultDownloader`

Reply:
The store can be used without an endpoint. There are many very valid use cases for this, e.g. you have a local store and want to do something without ever listening on the network.

I guess you could do a thing which combines a store and an endpoint, but then that is what the downloader *is*. Maybe there needs to be a thing that combines a store, endpoint and downloader.

I want to keep the downloader as a separate thing though since it has some internal state and there might be situations where you want to create/destroy them or even have multiple independent of the store.